pyomo/contrib/observer/model_observer.py

```diff
@@ -0,0 +1,1294 @@
+            self._active_constraints.pop(con)
+            self._named_expressions.pop(con, None)
+            self._external_functions.pop(con, None)
+            self._vars_referenced_by_con.pop(con)
+            self._params_referenced_by_con.pop(con)
+
+    def remove_constraints(self, cons: Collection[ConstraintData]):
+        self._remove_constraints(cons)
+        self._updates.run()
+
+    def _remove_sos_constraints(self, cons: Collection[SOSConstraintData]):
+        for con in cons:
+            if con not in self._active_sos:
+                raise ValueError(
+                    f'Cannot remove constraint {con.name} - it was not added'
+                )
+            self._updates.sos_to_update[con] |= Reason.removed
+            for v in self._vars_referenced_by_con[con]:
+                self._referenced_variables[v][1].remove(con)
+            for p in self._params_referenced_by_con[con]:
+                self._referenced_params[p][1].remove(con)
+            self._check_to_remove_vars(self._vars_referenced_by_con[con])
+            self._check_to_remove_params(self._params_referenced_by_con[con])
+            self._active_sos.pop(con)
+            self._vars_referenced_by_con.pop(con)
+            self._params_referenced_by_con.pop(con)
+
+    def remove_sos_constraints(self, cons: Collection[SOSConstraintData]):
+        self._remove_sos_constraints(cons)
+        self._updates.run()
+
+    def _remove_variables(self, variables: Collection[VarData]):
+        for v in variables:
+            if v not in self._referenced_variables:
+                raise ValueError(
+                    f'Cannot remove variable {v.name} - it has not been added'
+                )
+            self._updates.vars_to_update[v] |= Reason.removed
+            for p in self._params_referenced_by_var[v]:
+                self._referenced_params[p][3].remove(v)
+            self._check_to_remove_params(self._params_referenced_by_var[v])
+            self._params_referenced_by_var.pop(v)
+            if any(self._referenced_variables[v]):
+                raise ValueError(
+                    f'Cannot remove variable {v.name} - it is still being used by constraints/objectives'
+                )
+            self._referenced_variables.pop(v)
+            self._vars.pop(v)
+
+    def _remove_parameters(self, params: Collection[ParamData]):
+        for p in params:
+            if p not in self._referenced_params:
+                raise ValueError(
+                    f'Cannot remove parameter {p.name} - it has not been added'
+                )
+            self._updates.params_to_update[p] |= Reason.removed
+            if any(self._referenced_params[p]):
+                raise ValueError(
+                    f'Cannot remove parameter {p.name} - it is still being used by constraints/objectives'
+                )
+            self._referenced_params.pop(p)
+            self._params.pop(p)
+
+    def _update_var_bounds(self, v: VarData):
```

Comment:
Is `update` the right verb here?  Would `_record_var_bounds` be more descriptive?